File: apps/web/src/lib/transfer/progress.ts
export interface Progress {
  transferred: bigint;
  total: bigint;
  /** 0–1, or 0 before the total is known. */
  fraction: number;
  bytesPerSecond: number;
  /** Seconds remaining, or null while the rate is still meaningless. */
  etaSeconds: number | null;
}

const EMPTY: Progress = {
  transferred: 0n,
  total: 0n,
  fraction: 0,
  bytesPerSecond: 0,
  etaSeconds: null,
};

/**
 * Transfer rate and ETA.
 *
 * Rate is smoothed over a short trailing window rather than computed against
 * the whole transfer: an average since the start barely moves after the first
 * minute, so a stall would keep showing a healthy number. Over a 100 GB
 * transfer that difference is the whole value of the readout.
 */
export class ProgressMeter {
  private total = 0n;
  private transferred = 0n;
  private samples: { at: number; bytes: bigint }[] = [];

  private static readonly WINDOW_MS = 5_000;

  start(total: bigint): void {
    this.total = total;
    this.transferred = 0n;
    this.samples = [{ at: performance.now(), bytes: 0n }];
  }

  set(transferred: bigint): void {
    this.transferred = transferred;

    const now = performance.now();
    this.samples.push({ at: now, bytes: transferred });

    // Keep one sample older than the window so the span is always full-length.
    let drop = 0;
    while (drop + 1 < this.samples.length && now - this.samples[drop + 1]!.at > ProgressMeter.WINDOW_MS) {
      drop++;
    }
    if (drop > 0) this.samples.splice(0, drop);
  }

  snapshot(): Progress {
    if (this.total <= 0n) return EMPTY;

    const first = this.samples[0];
    const last = this.samples[this.samples.length - 1];

    let bytesPerSecond = 0;
    if (first && last && last.at > first.at) {
      bytesPerSecond = (Number(last.bytes - first.bytes) * 1000) / (last.at - first.at);
    }

    const remaining = Number(this.total - this.transferred);
    return {
      transferred: this.transferred,
      total: this.total,
      fraction: Number(this.transferred) / Number(this.total),
      bytesPerSecond,
      etaSeconds: bytesPerSecond > 0 && remaining > 0 ? remaining / bytesPerSecond : null,
    };
  }
}

File: apps/mobile/packages/direct_protocol/lib/src/progress.dart
/// Transfer rate and ETA.
///
/// Rate is smoothed over a short trailing window rather than averaged over
/// the whole transfer: an average barely moves after the first minute, so a
/// stall would keep showing a healthy number. Over a large transfer that
/// difference is the entire value of the readout.
///
/// Mirrors apps/web/src/lib/transfer/progress.ts.
class Progress {
  final int transferred;
  final int total;
  final double bytesPerSecond;
  final double? etaSeconds;

  /// Recent rate samples, oldest first, for the throughput sparkline.
  final List<double> history;

  /// Seconds since the transfer started.
  final double elapsedSeconds;

  const Progress({
    this.transferred = 0,
    this.total = 0,
    this.bytesPerSecond = 0,
    this.etaSeconds,
    this.history = const [],
    this.elapsedSeconds = 0,
  });

  double get fraction => total <= 0 ? 0 : transferred / total;
}

class _Sample {
  final int at;
  final int bytes;
  const _Sample(this.at, this.bytes);
}

class ProgressMeter {
  static const int _windowMs = 5000;
  static const int _tickMs = 500;
  static const int _historyLength = 48;

  final Stopwatch _clock = Stopwatch();
  final List<_Sample> _samples = [];

  /// One rate reading per tick, capped — the sparkline's data.
  final List<double> _rates = [];
  int _lastTick = 0;

  int _total = 0;
  int _transferred = 0;

  void start(int total) {
    _total = total;
    _transferred = 0;
    _lastTick = 0;
    _samples
      ..clear()
      ..add(const _Sample(0, 0));
    _rates.clear();
    _clock
      ..reset()
      ..start();
  }

  void set(int transferred) {
    _transferred = transferred;
    final now = _clock.elapsedMilliseconds;
    _samples.add(_Sample(now, transferred));

    // Keep one sample older than the window so the span stays full-length.
    var drop = 0;
    while (drop + 1 < _samples.length &&
        now - _samples[drop + 1].at > _windowMs) {
      drop++;
    }
    if (drop > 0) _samples.removeRange(0, drop);

    // Sample the smoothed rate on a fixed cadence, so the sparkline's x axis
    // is time rather than 'however often chunks happened to land'.
    if (now - _lastTick >= _tickMs) {
      _lastTick = now;
      _rates.add(_rate());
      if (_rates.length > _historyLength) _rates.removeAt(0);
    }
  }

  double _rate() {
    final first = _samples.first;
    final last = _samples.last;
    if (last.at <= first.at) return 0;
    return (last.bytes - first.bytes) * 1000 / (last.at - first.at);
  }

  Progress snapshot() {
    if (_total <= 0) return const Progress();

    final rate = _rate();
    final remaining = _total - _transferred;

    return Progress(
      transferred: _transferred,
      total: _total,
      bytesPerSecond: rate,
      etaSeconds: rate > 0 && remaining > 0 ? remaining / rate : null,
      history: List.unmodifiable(_rates),
      elapsedSeconds: _clock.elapsedMilliseconds / 1000,
    );
  }
}
